### PR TITLE
complete 'optimization' and declutter and maybe mostly rewrite

### PR DESCRIPTION
**Pull request by wael444**
+ changed script to bash to avoid certain shell language errors, or to be safe and avoid EOF in for checking files.
+ as per that `glow` is recommended, everything else is removed. it can be changed via the `MDRENDERER` variable.
+ optimized `has_command()`
+ removed `SKIP_OK` as this doesn't make sense in my opinion, why would the user see what is okay? they would probably want what to change.
+ checking for XDG directories is extremely inefficient. it has many duplicate lines  and should have been entirely shortened to a function. it is generally known that whenever you have duplicate lines, you can almost **ALWAYS** shorten it to a function. exactly as i did with this and `logsc`.
+ `apply_shell_expansion()` is obseleted by `eval echo` + `decode_string()` 
+ `decode_string` optimized to use bash substitution, which makes it faster, also' normalize number of trailing newlines to 2' doesn't make much sense to me so i removed it.
+ removal of `check_file` INFO and SUCS as `SKIP_OK` has been removed.
+ removal of help, as `SKIP_OK` has been removed.
     + also, please dont use a for loop and if to check for arguments, as that is an extremely shitty way of doing it use `case` instead. i can show you what i mean if you need `SKIP_OK` back.
+ removal of `do_check_programs()` and `check_programs` as they are obsoleted by simply having them there.
+ removal of `IFS` as they don't change much.
+ removal of comments
+ `dirname $0` is unreccomended, see my implementation for more details.

+ the output is EXACTLY the same.

if you have ANY problems with my PR please let me know, i would be happy to add whatever i had removed back.

### CODE REVIEW

**Comment by b3nj5m1n** [comment]
I have to ask. Do you seriously think I'm going to merge this, or is this a troll?

**Comment by wael444** [comment]
I put effort and time into this PR. Of course I'm not trolling.

If you don't like my optimization(s) you can tell me what is wrong with it.

I simply want to help, as I want to use this program.

**Comment by b3nj5m1n** [comment]
> changed script to bash to avoid certain shell language errors, or to be safe and avoid EOF in for checking files.

Me and several other people (@Ballasi, @Mudskipper875, @2641a40fd44383320adde4b027a1d0b03bd550) have put quite a bit of work into making this script posix compliant, see #11, #27, #6, and #16.

>  as per that glow is recommended, everything else is removed. it can be changed via the MDRENDERER variable.

Glow is recommended, and there are several fallbacks in case someone doesn't have/want glow. Why would you remove those? See #69.

> optimized has_command()

It's literally the same thing, except that your version exits, so it's a different function.

> removed SKIP_OK as this doesn't make sense in my opinion, why would the user see what is okay? they would probably want what to change.

Which is why it's hidden behind the verbose flag.

> checking for XDG directories is extremely inefficient. it has many duplicate lines and should have been entirely shortened to a function. it is generally known that whenever you have duplicate lines, you can almost ALWAYS shorten it to a function. exactly as i did with this and logsc.

It's not inefficient, it's literally the same as what you're doing. This is likely never going to change, so imo there's no point in abstracting it out into a function.

>  apply_shell_expansion() is obseleted by eval echo + decode_string()

This is unsafe.

> decode_string optimized to use bash substitution, which makes it faster, also' normalize number of trailing newlines to 2' doesn't make much sense to me so i removed it.

- We can't use anything bash specific
- #70

> removal of check_file INFO and SUCS as SKIP_OK has been removed.

Idk how to respond to that, I didn't implement this to be removed?

> removal of help, as SKIP_OK has been removed.

Even if there are no flags, you should still have a help command that tells you what the program does. And we're not removing -v, so there are flags.

> also, please dont use a for loop and if to check for arguments, as that is an extremely shitty way of doing it use case instead. i can show you what i mean if you need SKIP_OK back.

There's nothing wrong with using if, I also wanted to rework the argument parsing to be a bit more sophisticated, so I haven't bothered to change anything about the current system yet.

> removal of do_check_programs() and check_programs as they are obsoleted by simply having them there.

This is bad practice.

>  removal of IFS as they don't change much.

I don't think that's posix compliant.

> removal of comments



>  dirname $0 is unreccomended, see my implementation for more details.

I don't think this is posix compliant.

> the output is EXACTLY the same.

It's not, you literally removed everything aside from the very base functionality.

---

> I simply want to help, as I want to use this program.

I appreciate you wanting to help, but I don't understand what's keeping you from using the program.

**Comment by wael444** [comment]
> Me and several other people (\@Ballasi,\@Mudskipper875, \@2641a40fd44383320adde4b027a1d0b03bd550) have put quite a bit of work into making this script posix compliant, see https://github.com/b3nj5m1n/xdg-ninja/issues/11, https://github.com/b3nj5m1n/xdg-ninja/pull/27, https://github.com/b3nj5m1n/xdg-ninja/pull/6, and https://github.com/b3nj5m1n/xdg-ninja/pull/16.

You should have told me that from the start.. i would have made the appropriate changes. 
but to be considerate, what makes a program POSIX compliant??

> Glow is recommended, and there are several fallbacks in case someone doesn't have/want glow. Why would you remove those? See https://github.com/b3nj5m1n/xdg-ninja/pull/69.

> Which is why it's hidden behind the verbose flag.

> decode_string optimized to use bash substitution, which makes it faster, also' normalize number of trailing newlines to 2' doesn't make much sense to me so i removed it.

> It's not, you literally removed everything aside from the very base functionality.

Like i said, you can just simply tell me to have the features back..

> It's literally the same thing, except that your version exits, so it's a different function

the message is built in, so that it can be used to searching the primary markdown renderer.

> It's not inefficient, it's literally the same as what you're doing. This is likely never going to change, so imo there's no point in abstracting it out into a function.

what?? are you telling me unclean code is better?

>  This is unsafe.

i personally do not have a clear understanding of `eval`, but how is your solution which uses `eval` as well safe?

> This is bad practice.

How? you simply told me it is without elaborating.

> I don't think that's posix compliant.

> I don't think this is posix compliant.

isn't the output the same either way? what changes?

> removal of comments

to put it simply, i didn't have time to comment my scripting, as i thought it was easy to understand.

I don't see why my PR was rejected still.

**Comment by wael444** [comment]
And to be honest, i have quite literally trimmed ton's of lines, could be probably atleast 131 SLOC if i added the features you wanted back.

but making it POSIX compliant might make it larger, albeit simpler.

**Comment by b3nj5m1n** [comment]
> but to be considerate, what makes a program POSIX compliant??

https://en.wikipedia.org/wiki/POSIX

Also see the issues I linked.

> what?? are you telling me unclean code is better?

Repetition ≠ unclean code.

> i personally do not have a clear understanding of eval, but how is your solution which uses eval as well safe?

Safe wasn't the right word, your solution doesn't handle all cases correctly, they're both not safe. See the source for that function for details.

> isn't the output the same either way? what changes?

It won't run on posix compliant shells, only bash.

**Comment by Ballasi** [comment]
Hi, putting my two cents here as I got pinged:

There are things I really like about the changes you've made, I personally find the markdown rendering check part or XDG environment variable part could be written differently to make it more readable and accessible, and the way you've tackled XDG env variable check is close to how I would've done it. But as b3nj5m1n, there are also changes (quite a few) I am not agreeing with.

The first, and main problem with your PR is that all of the changes are regrouped in one commit, and there are so many changes, which makes it hardly reviewable. There might have been changes b3nj5m1n would agree on but have been ignored because they have been overshadowed by the things he pointed out, and that's unfortunate.

If you want to have more luck with PRs, including here, it's important to make the review easy and accessible. Personally, "optimize root" didn't help me much understand what changes and the PR message doesn't point out where the changes are done. Splitting in different commits could help that and would make it easier to remove changes that are deemed unnecessary.

On the topic of POSIX-compliancy, it includes the downside of "more work"/"more SLOC" as stated in corresponding issues/PR but with the advantage of being available on literally all Linux shells (except shells like `fish`, which basically is their main motive). I personally find it largely bearable in small programs such as this one so I wouldn't call "trimming tons of lines" a boon.

I'd also like to note that the way b3nj5m1n answered wasn't meant --I hope!-- to be harmful towards you. The changes were refused not because they were deemed bad but because it wasn't a fit with the program's design goal. I hope it is clear on both sides that the other has good intentions.

**Comment by wael444** [comment]
> The first, and main problem with your PR is that all of the changes are regrouped in one commit, and there are so many changes, which makes it hardly reviewable. There might have been changes b3nj5m1n would agree on but have been ignored because they have been overshadowed by the things he pointed out, and that's unfortunate.

> If you want to have more luck with PRs, including here, it's important to make the review easy and accessible. Personally, "optimize root" didn't help me much understand what changes and the PR message doesn't point out where the changes are done. Splitting in different commits could help that and would make it easier to remove changes that are deemed unnecessary.

i did this to have one entire clean commit since having too many commits is unclean.
I'm also not very good at making things very clear, as i had made tons of changes i can't really document very clearly.
if you both want me to split the changes I'd be more than happy to help.

> On the topic of POSIX-compliancy, it includes the downside of "more work"/"more SLOC" 

why does POSIX compliancy matter? if the script is under bash then it should work on all systems, on most systems `sh` is symlinked to many shells in many operating systems.

> 'd also like to note that the way b3nj5m1n answered wasn't meant --I hope!-- to be harmful towards you.

well to be honest, it was kind of mean :(, but i can see why in a way.

**Comment by Ballasi** [comment]
> i did this to have one entire clean commit since having too many commits is unclean

To be honest, this is considered unclean by git's official guidelines (see https://git-scm.com/docs/SubmittingPatches#separate-commits) and I'd consider it unclean to have one large commit as this is not what they were meant to be in the first place (cc. NVIDIA's kernel modules, doing a bad job at it)

> why does POSIX compliancy matter? if the script is under bash then it should work on all systems

While this may be true for most --if not all-- desktop-focused Linux distributions, this isn't necessarily the case for any other systems, especially embedded devices, servers or docker containers which are not likely to package "heavy"/"useless" (in their context) programs such as bash.

However, you can say the program's usage may not likely happen in these contexts, that's why when I first talked with b3nj5m1n about POSIX-compliancy, it was mostly just an idea and it wasn't much mandatory. But it also wasn't that hard to implement, and is a nice feature to have (you never know, if someone hates bash or if you have a CI running a docker container and you have to check the correctness of the home directory, it's all working with POSIX-compliancy!). That's why I left him the choice as the program was initially only working with bash.

Basically, it doesn't matter much, but it was a nice-to-have feature and easy to implement, so why not have it?

> well to be honest, it was kind of mean :(

That's why I wanted to clarify that it wasn't the initial intention.

**Comment by wael444** [comment]
> Basically, it doesn't matter much, but it was a nice-to-have feature and easy to implement, so why not have it?

can you tell me what is not POSIX Compliant in my modified version of xdg-ninja?

and just as simple question, should this PR even be noticed? i mean that should it even be re-opened in a new PR? or just re-opened here with new commits describing each change?

I don't know exactly how to do each change but i think i can figure it out.

**Comment by b3nj5m1n** [comment]
Thanks a lot @Ballasi for chiming in and answering some of the questions.

> can you tell me what is not POSIX Compliant in my modified version of xdg-ninja?

When you have `#!/usr/bin/env sh` at the top of the script and run `shellcheck xdg-ninja.sh`, it should tell you what's not posix compliant.

> There are things I really like about the changes you've made, I personally find the markdown rendering check part or XDG environment variable part could be written differently to make it more readable and accessible, and the way you've tackled XDG env variable check is close to how I would've done it.

These are the things I'd be more willing to accept, even though I don't think there's a real problem with either.

> and just as simple question, should this PR even be noticed? i mean that should it even be re-opened in a new PR? or just re-opened here with new commits describing each change?

Open a new PR for each one of those two changes.

> well to be honest, it was kind of mean :(

I'm sorry, I didn't mean to be insulting.

**Comment by Ballasi** [comment]
Thanks @b3nj5m1n for the clarification! I'd like to add a few things:

> can you tell me what is not POSIX Compliant in my modified version of xdg-ninja?

As b3nj5m1n said, ShellCheck is a good tool to point out POSIX-compliancy issue. In this repository, there's a GitHub actions CI to run it. However, in your case, it doesn't show any related issues as you have defined `#!/usr/bin/env bash` at the beginning of your program. (Sorry for the redundancy, I wrote this before b3nj5m1n answered!)

When running ShellCheck, I find these issues:

- line 15 and 16: string replacement is a bashism (see [SC3060](https://www.shellcheck.net/wiki/SC3060))
- there are also here-strings in line 73 that we used to have. I remember have been fixed in its specific PR (see [SC3011](https://www.shellcheck.net/wiki/SC3011))

Overall, I'd also recommend you to run ShellCheck and have no issue before making any commit. This helps avoid any possible exploits (word splitting) or incorrect code (I see you have A && B || C patterns which probably isn't behaving the way you want it to be).

> I don't know exactly how to do each change but i think i can figure it out.

You can probably just do `git reset HEAD^` so you go back one commit while keeping your changes unstaged. From there you can stage the lines you want and work the way up to have the whole file committed (I have no idea how to stage specific lines on git files using the CLI, I usually use the `git gui` tool to do this specifically for me).

**Comment by wael444** [comment]
> Open a new PR for each one of those two changes.

i will try to open a PR that only really makes functions the code, making it cleaner, but hopefully still being the same. then another to modify the actual program itself.

but just a few questions,
+ should the user use `$MDRENDERER` or to have the script automatically check for other markdown renderers?

> Overall, I'd also recommend you to run ShellCheck

thanks for introducing me to ShellCheck, i have never heard of it before.

> This helps avoid any possible exploits (word splitting) or incorrect code (I see you have A && B || C patterns which probably isn't behaving the way you want it to be).

i wouldn't say that i know how it works but i'm sure that it should theoretically work.

**Comment by wael444** [comment]
Also, what is `unalias -a` for?

**Comment by Ballasi** [comment]
> Also, what is `unalias -a` for?

From `man unalias`:

```
NAME
       unalias — remove alias definitions

DESCRIPTION
       The  unalias  utility shall remove the definition for each alias name specified. See
       Section 2.3.1, Alias Substitution.  The aliases shall be removed  from  the  current
       shell execution environment; see Section 2.12, Shell Execution Environment.

OPTIONS
       -a        Remove all alias definitions from the current shell execution environment.

RATIONALE
       The -a option is the equivalent of the unalias * form of the C shell and is provided
       to address security concerns about unknown aliases entering  the  environment  of  a
       user  (or application) through the allowable implementation-defined predefined alias
       route or as a result of an ENV file.
```

**Comment by wael444** [comment]
What purpose does it serve in the script for?

**Comment by Ballasi** [comment]
It disables any aliases the user has set.

If the user has set any aliases for commands used in the programs, without `unalias -a`, it would just run the alias rather than the actual command, this isn't intended and `unalias` helps disable that.
